File: Shortest Path/Dijkstra/2151번 - 거울 설치/2151.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define X first
#define Y second
#define INF 1e9

int n;
char home[51][51];
int vis[51][51][4] = { 0, };
vector<pair<int, int> > door;

int dx[4] = { 0, -1, 0, 1 };
int dy[4] = { -1, 0, 1, 0 };

void input() {
    cin >> n;
    for(int i = 0; i < n; i++) {
        string s;
        cin >> s;
        for(int j = 0; j < n; j++) {
            home[i][j] = s[j];
            if(home[i][j] == '#') door.push_back({i, j});
        }
    }
    
    fill_n(&vis[0][0][0], 51 * 51 * 4, INF);
}

void solve() {
    // 시작 점에서 탐색 4방향
    priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>> > pq;
    for(int i = 0; i < 4; i++) pq.push({0, i, door[0].X, door[0].Y});
    
    while(!pq.empty()) {
        int cnt = pq.top()[0];
        int d = pq.top()[1];
        int x = pq.top()[2], y = pq.top()[3];
        pq.pop();
        
        if(x == door[1].X && y == door[1].Y) {
            cout << cnt << '\n';
            break;
        }
        
        int nx = x + dx[d];
        int ny = y + dy[d];
        
        if(nx < 0 || nx >= n || ny < 0 || ny >= n || home[nx][ny] == '*') continue;
        
        // 다음 칸의 현재 방향대로 거울 개수가 더 낮을 때
        if(vis[nx][ny][d] > cnt) {
            vis[nx][ny][d] = cnt;
            pq.push({cnt, d, nx, ny});
        }
        
        // 다음 칸에서 거울을 설치할 수 있는 경우
        if(home[nx][ny] == '!') {
            int ld = (d + 3) % 4, rd = (d + 1) % 4;
            
            // 거울 설치할 수 있는 칸에서 좌측으로 갈 때 거울 개수가 더 낮을 때
            if(vis[nx][ny][ld] > cnt + 1) {
                vis[nx][ny][ld] = cnt + 1;
                pq.push({cnt + 1, ld, nx, ny});
            }
            // 거울 설치할 수 있는 칸에서 우측으로 갈 때 거울 개수가 더 낮을 때
            if(vis[nx][ny][rd] > cnt + 1) {
                vis[nx][ny][rd] = cnt + 1;
                pq.push({cnt + 1, rd, nx, ny});
            }
        }
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Shortest Path/Dijkstra/2151번 - 거울 설치/2151.py
import sys
from collections import deque
import heapq

input = sys.stdin.readline

INF = float('inf')

n = int(input())
home = [input().rstrip() for _ in range(n)]
door = []
for i in range(n):
    for j in range(n):
        if home[i][j] == '#':
            door.append((i, j))

def solve():
    # 각 칸마다 4방향의 거울 개수 기록
    vis = [[[INF] * 4 for _ in range(n)] for _ in range(n)]
    hq = []
    for i in range(4):
        heapq.heappush(hq, (0, door[0], i))
    
    dx = [0, -1, 0, 1]
    dy = [-1, 0, 1, 0]
    
    while hq:
        cnt, (coor_x, coor_y), d = heapq.heappop(hq)
        
        nx = coor_x + dx[d]
        ny = coor_y + dy[d]
        
        # 가장 적은 거울 개수일 때의 다른 문에 도착 시
        if door[1] == (nx, ny):
            print(cnt)
            break
        
        if nx < 0 or nx >= n or ny < 0 or ny >= n or home[nx][ny] == '*':
            continue
        
        # 다음 방문 칸의 방향의 거울 개수를 더 작게 가능한 경우
        if vis[nx][ny][d] > cnt:
            vis[nx][ny][d] = cnt
            heapq.heappush(hq, (cnt, (nx, ny), d))
            
        # 다음 방문 칸에 거울을 설치가 가능한 경우
        if home[nx][ny] == '!':
            ld, rd = (d + 3) % 4, (d + 1) % 4
            
            # 거울을 좌측 방향으로 했을 때
            if vis[nx][ny][ld] > cnt + 1:
                vis[nx][ny][ld] = cnt + 1
                heapq.heappush(hq, (cnt + 1, (nx, ny), ld))
                
            # 거울을 우측 방향으로 했을 때
            if vis[nx][ny][rd] > cnt + 1:
                vis[nx][ny][rd] = cnt + 1
                heapq.heappush(hq, (cnt + 1, (nx, ny), rd))


if __name__ == '__main__':
    solve()